cpp/src/stream_compaction/distinct_helpers.cu

```diff
@@ -23,17 +23,19 @@ namespace cudf::detail {
 
 template <typename RowHasher>
 rmm::device_uvector<size_type> reduce_by_row(hash_set_type<RowHasher>& set,
-                                             size_type set_size,
                                              size_type num_rows,
                                              duplicate_keep_option keep,
                                              rmm::cuda_stream_view stream,
                                              rmm::device_async_resource_ref mr)
 {
-  auto output_indices = rmm::device_uvector<size_type>(set_size, stream, mr);
+  auto output_indices = rmm::device_uvector<size_type>(num_rows, stream, mr);
 
   // If we don't care about order, just gather indices of distinct keys taken from set.
   if (keep == duplicate_keep_option::KEEP_ANY) {
+    auto const iter = thrust::counting_iterator<cudf::size_type>{0};
+    auto const size = set.insert(iter, iter + num_rows, stream.value());
     set.retrieve_all(output_indices.begin(), stream.value());
+    output_indices.resize(size, stream);
```

Comment:
Sorry for not being clear, I was suggesting something like this:
```cuda
    auto const iter = thrust::counting_iterator<cudf::size_type>{0};
    set.insert_async(iter, iter + num_rows, stream.value());
    auto const output_end = set.retrieve_all(output_indices.begin(), stream.value());
    output_indices.resize(std::distance(output_indices.begin(), output_end), stream);
```
Using the asynchronous insert is more efficient since it doesn't have the atomic counter counting the number of successful insertions.

Reply:
Got it!. That indeed improved the performance. Thank you!!